pkg/buildid/buildid.go

```diff
@@ -56,29 +56,84 @@ func BuildID(path string) (string, error) {
 	}
 	f.Close()
 
+	if id, err := fastGNUBuildID(path); err == nil && len(id) > 0 {
+		return hex.EncodeToString(id), nil
+	}
+
 	return elfBuildID(path)
 }
 
-func fastGoBuildID(path string) (string, error) {
-	elfFile, err := elf.Open(path)
+func fastGoBuildID(path string) ([]byte, error) {
+	findBuildID := func(notes []elfreader.ElfNote) ([]byte, error) {
+		var buildID []byte
+		for _, note := range notes {
+			if note.Name == "Go" && note.Type == elfreader.NoteTypeGoBuildID {
+				if buildID == nil {
+					buildID = note.Desc
+				} else {
+					return nil, fmt.Errorf("multiple build ids found, don't know which to use")
+				}
+			}
+		}
+		return buildID, nil
+	}
+	data, err := extractNote(path, ".note.go.buildid", findBuildID)
 	if err != nil {
-		return "", fmt.Errorf("failed to open elf: %w", err)
+		return nil, err
 	}
-	defer elfFile.Close()
+	return data, nil
+}
 
-	s := elfFile.Section(".note.go.buildid")
+func fastGNUBuildID(path string) ([]byte, error) {
+	findBuildID := func(notes []elfreader.ElfNote) ([]byte, error) {
+		var buildID []byte
+		for _, note := range notes {
+			if note.Name == "GNU" && note.Type == elfreader.NoteTypeGNUBuildID {
+				if buildID == nil {
+					buildID = note.Desc
+				} else {
+					return nil, fmt.Errorf("multiple build ids found, don't know which to use")
+				}
+			}
+		}
+		return buildID, nil
+	}
+	data, err := extractNote(path, ".note.gnu.build-id", findBuildID)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+func extractNote(path, section string, findBuildID func(notes []elfreader.ElfNote) ([]byte, error)) ([]byte, error) {
+	f, err := elf.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open elf: %w", err)
+	}
+	defer f.Close()
+
+	s := f.Section(section)
 	if s == nil {
-		return "", fmt.Errorf("failed to find .note.go.buildid section")
+		return nil, fmt.Errorf("failed to find %s section", section)
 	}
+
+	notes, err := elfreader.ParseNotes(s.Open(), int(s.Addralign), f.ByteOrder)
```

Comment:
Do we have to close it?

Reply:
Noop, we don't. Section is not a Closer.

Reply:
Oops misread and thought it was returning a ReadWriteCloser rather than a ReadSeeker, my bad!